File: sdk/typescript/examples/analytics-demo.ts
/**
 * Analytics and Budget Management Demo
 *
 * This example demonstrates the Circuit Breaker SDK's analytics and budget management
 * capabilities, showing how to track costs, set budgets, and monitor spending across
 * different users and projects.
 */

import { Client } from '../src/client.js';
import {
  budgetStatus,
  costAnalytics,
  setBudget,
  getUserBudgetStatus,
  setUserMonthlyBudget,
  type BudgetStatus,
  type CostAnalytics,
} from '../src/analytics.js';

async function main(): Promise<void> {
  console.log('🔍 Circuit Breaker Analytics & Budget Management Demo');
  console.log('====================================================');

  // Initialize the client
  const baseUrl = process.env.CIRCUIT_BREAKER_URL || 'http://localhost:4000';
  const apiKey = process.env.CIRCUIT_BREAKER_API_KEY;

  let client = Client.builder().baseUrl(baseUrl);
  if (apiKey) {
    client = client.apiKey(apiKey);
  }
  const circuitBreakerClient = client.build();

  // Test connection
  try {
    const ping = await circuitBreakerClient.ping();
    console.log(`✅ Connected to Circuit Breaker server: ${ping.message}`);
  } catch (error) {
    console.log(`❌ Failed to connect to server: ${error}`);
    console.log(
      `   Make sure the Circuit Breaker server is running at ${baseUrl}`,
    );
    return;
  }

  console.log('\n📊 Analytics & Budget Management Features:');
  console.log('==========================================');

  // 1. Budget Management Demo
  console.log('\n1. 💰 Budget Management');
  console.log('   ---------------------');

  const userId = 'demo_user_123';
  const projectId = 'demo_project_456';

  // Set a monthly budget for a user
  console.log(`   Setting monthly budget for user: ${userId}`);
  try {
    const budget = await circuitBreakerClient
      .analytics()
      .setBudget()
      .userId(userId)
      .limit(100.0)
      .period('monthly')
      .warningThreshold(0.8)
      .execute();

    console.log('   ✅ Budget set successfully:');
    console.log(`      • Budget ID: ${budget.budgetId}`);
    console.log(`      • Limit: $${budget.limit.toFixed(2)}`);
    console.log(
      `      • Used: $${budget.used.toFixed(2)} (${budget.percentageUsed.toFixed(1)}%)`,
    );
    console.log(`      • Remaining: $${budget.remaining.toFixed(2)}`);
    console.log(`      • Status: ${budget.message}`);

    if (budget.isWarning) {
      console.log('      ⚠️  Warning: Budget is in warning state!');
    }
    if (budget.isExhausted) {
      console.log('      🚨 Alert: Budget is exhausted!');
    }
  } catch (error) {
    console.log(
      `   ⚠️  Could not set budget (server may not be running): ${error}`,
    );
  }

  // Set a project budget
  console.log(`\n   Setting project budget for: ${projectId}`);
  try {
    const budget = await circuitBreakerClient
      .analytics()
      .setBudget()
      .projectId(projectId)
      .limit(500.0)
      .period('monthly')
      .warningThreshold(0.75)
      .execute();

    console.log('   ✅ Project budget set:');
    console.log(`      • Limit: $${budget.limit.toFixed(2)}`);
    console.log(`      • Warning at: ${(0.75 * 100).toFixed(0)}%`);
  } catch (error) {
    console.log(`   ⚠️  Could not set project budget: ${error}`);
  }

  // 2. Budget Status Monitoring
  console.log('\n2. 📈 Budget Status Monitoring');
  console.log('   ----------------------------');

  // Check user budget status
  console.log(`   Checking budget status for user: ${userId}`);
  try {
    const status = await circuitBreakerClient
      .analytics()
      .budgetStatus()
      .userId(userId)
      .get();

    console.log('   ✅ User budget status:');
    displayBudgetStatus(status);
  } catch (error) {
    console.log(`   ⚠️  Could not get user budget status: ${error}`);
  }

  // Check project budget status
  console.log(`\n   Checking budget status for project: ${projectId}`);
  try {
    const status = await circuitBreakerClient
      .analytics()
      .budgetStatus()
      .projectId(projectId)
      .get();

    console.log('   ✅ Project budget status:');
    displayBudgetStatus(status);
  } catch (error) {
    console.log(`   ⚠️  Could not get project budget status: ${error}`);
  }

  // 3. Cost Analytics
  console.log('\n3. 📊 Cost Analytics');
  console.log('   -----------------');

  const startDate = '2024-01-01';
  const endDate = '2024-01-31';

  console.log(
    `   Getting cost analytics for user: ${userId} (${startDate} to ${endDate})`,
  );
  try {
    const analytics = await circuitBreakerClient
      .analytics()
      .costAnalytics()
      .userId(userId)
      .dateRange(startDate, endDate)
      .get();

    console.log('   ✅ Cost analytics retrieved:');
    displayCostAnalytics(analytics);
  } catch (error) {
    console.log(`   ⚠️  Could not get cost analytics: ${error}`);
  }

  // Get project analytics
  console.log(`\n   Getting cost analytics for project: ${projectId}`);
  try {
    const analytics = await circuitBreakerClient
      .analytics()
      .costAnalytics()
      .projectId(projectId)
      .dateRange(startDate, endDate)
      .get();

    console.log('   ✅ Project cost analytics:');
    displayCostAnalytics(analytics);
  } catch (error) {
    console.log(`   ⚠️  Could not get project analytics: ${error}`);
  }

  // 4. Convenience Functions Demo
  console.log('\n4. 🛠️  Convenience Functions');
  console.log('   -------------------------');

  // Using convenience functions
  console.log('   Using convenience functions for common operations:');

  // Budget status convenience function
  try {
    const status = await budgetStatus(circuitBreakerClient)
      .userId(userId)
      .get();

    console.log(
      `   ✅ Convenience budget status: $${status.used.toFixed(2)} used of $${status.limit.toFixed(2)}`,
    );
  } catch (error) {
    console.log(`   ⚠️  Convenience budget status failed: ${error}`);
  }

  // Cost analytics convenience function
  try {
    const analytics = await costAnalytics(
      circuitBreakerClient,
      startDate,
      endDate,
    )
      .userId(userId)
      .get();

    console.log(
      `   ✅ Convenience analytics: $${analytics.totalCost.toFixed(2)} total cost`,
    );
  } catch (error) {
    console.log(`   ⚠️  Convenience analytics failed: ${error}`);
  }

  // Set budget convenience function
  try {
    const budget = await setBudget(circuitBreakerClient, 200.0, 'monthly')
      .userId('convenience_user')
      .execute();

    console.log(`   ✅ Convenience budget set: $${budget.limit.toFixed(2)} limit`);
  } catch (error) {
    console.log(`   ⚠️  Convenience budget set failed: ${error}`);
  }

  // High-level convenience functions
  try {
    const userBudget = await getUserBudgetStatus(circuitBreakerClient, userId);
    console.log(
      `   ✅ User budget convenience: ${userBudget.percentageUsed.toFixed(1)}% used`,
    );
  } catch (error) {
    console.log(`   ⚠️  User budget convenience failed: ${error}`);
  }

  try {
    const monthlyBudget = await setUserMonthlyBudget(
      circuitBreakerClient,
      'monthly_user',
      150.0,
      0.9,
    );
    console.log(
      `   ✅ Monthly budget convenience: $${monthlyBudget.limit.toFixed(2)} with 90% warning`,
    );
  } catch (error) {
    console.log(`   ⚠️  Monthly budget convenience failed: ${error}`);
  }

  // 5. Real-time Cost Monitoring (Future Feature)
  console.log('\n5. ⏰ Real-time Cost Monitoring');
  console.log('   -----------------------------');
  console.log(
    '   Real-time cost monitoring via subscriptions will be available in a future release.',
  );
  console.log('   This will allow you to:');
  console.log('   • Subscribe to cost updates as they happen');
  console.log('   • Get real-time alerts when budgets are exceeded');
  console.log('   • Monitor spending patterns in real-time');

  // Demonstrate that subscription isn't implemented yet
  try {
    await circuitBreakerClient.analytics().subscribeCostUpdates(userId);
    console.log('   ✅ Subscribed to cost updates');
  } catch (error) {
    console.log(`   ⚠️  Cost update subscriptions: ${error}`);
  }

  // 6. Advanced Analytics Scenarios
  console.log('\n6. 🔬 Advanced Analytics Scenarios');
  console.log('   --------------------------------');

  // Multi-date range analysis
  const dateRanges = [
    { start: '2024-01-01', end: '2024-01-31', month: 'January' },
    { start: '2024-02-01', end: '2024-02-29', month: 'February' },
    { start: '2024-03-01', end: '2024-03-31', month: 'March' },
  ];

  console.log('   Analyzing costs across multiple months:');
  for (const { start, end, month } of dateRanges) {
    try {
      const analytics = await circuitBreakerClient
        .analytics()
        .costAnalytics()
        .userId(userId)
        .dateRange(start, end)
        .get();

      console.log(
        `   • ${month}: $${analytics.totalCost.toFixed(2)} total, ${analytics.totalTokens} tokens used`,
      );
    } catch (error) {
      console.log(`   • ${month}: No data available`);
    }
  }

  // Budget health check
  console.log('\n   Budget Health Check:');
  try {
    const status = await circuitBreakerClient
      .analytics()
      .budgetStatus()
      .userId(userId)
      .get();

    let health: string;
    if (status.isExhausted) {
      health = '🚨 CRITICAL';
    } else if (status.isWarning) {
      health = '⚠️  WARNING';
    } else if (status.percentageUsed > 50.0) {
      health = '🟡 MODERATE';
    } else {
      health = '✅ HEALTHY';
    }

    console.log(
      `   Budget Health: ${health} (${status.percentageUsed.toFixed(1)}% used)`,
    );

    // Recommendations
    if (status.isExhausted) {
      console.log('   💡 Recommendation: Increase budget limit or optimize usage');
    } else if (status.isWarning) {
      console.log('   💡 Recommendation: Monitor usage closely, consider optimizations');
    } else if (status.percentageUsed > 50.0) {
      console.log(
        '   💡 Recommendation: Review usage patterns for optimization opportunities',
      );
    } else {
      console.log('   💡 Budget is healthy - continue current usage patterns');
    }
  } catch (error) {
    console.log(`   ⚠️  Could not perform health check: ${error}`);
  }

  // 7. TypeScript-specific Features
  console.log('\n7. 🔷 TypeScript-Specific Features');
  console.log('   --------------------------------');

  console.log('   Type-safe analytics operations:');

  // Type-safe budget operations
  try {
    const budget = await circuitBreakerClient
      .analytics()
      .setBudget()
      .userId('typescript_user')
      .limit(75.0)
      .period('weekly')
      .warningThreshold(0.6)
      .execute();

    // TypeScript ensures type safety
    const isOverBudget: boolean = budget.isExhausted;
    const remainingBudget: number = budget.remaining;
    const budgetMessage: string = budget.message;

    console.log('   ✅ Type-safe budget operations:');
    console.log(`      • Over budget: ${isOverBudget}`);
    console.log(`      • Remaining: $${remainingBudget.toFixed(2)}`);
    console.log(`      • Status: ${budgetMessage}`);
  } catch (error) {
    console.log(`   ⚠️  Type-safe budget operations failed: ${error}`);
  }

  // Type-safe analytics with destructuring
  try {
    const {
      totalCost,
      totalTokens,
      averageCostPerToken,
      providerBreakdown,
      modelBreakdown,
    } = await circuitBreakerClient
      .analytics()
      .costAnalytics()
      .userId('typescript_user')
      .dateRange('2024-01-01', '2024-01-31')
      .get();

    console.log('   ✅ Type-safe analytics with destructuring:');
    console.log(`      • Total cost: $${totalCost.toFixed(2)}`);
    console.log(`      • Total tokens: ${totalTokens}`);
    console.log(`      • Avg cost/token: $${averageCostPerToken.toFixed(6)}`);
    console.log(`      • Provider breakdown entries: ${Object.keys(providerBreakdown).length}`);
    console.log(`      • Model breakdown entries: ${Object.keys(modelBreakdown).length}`);
  } catch (error) {
    console.log(`   ⚠️  Type-safe analytics failed: ${error}`);
  }

  console.log('\n🎉 Analytics Demo Complete!');
  console.log('============================');
  console.log('This demo showcased:');
  console.log('• Budget management for users and projects');
  console.log('• Real-time budget status monitoring');
  console.log('• Comprehensive cost analytics');
  console.log('• Convenience functions for common operations');
  console.log('• Advanced analytics scenarios and health checks');
  console.log('• TypeScript-specific type safety features');
  console.log('\nThe Analytics client provides powerful tools for:');
  console.log('• Cost control and budget management');
  console.log('• Usage optimization and monitoring');
  console.log('• Financial planning and reporting');
  console.log('• Multi-tenant cost tracking');
  console.log('• Type-safe analytics operations');
}

function displayBudgetStatus(status: BudgetStatus): void {
  console.log(`      • Budget ID: ${status.budgetId}`);
  console.log(`      • Limit: $${status.limit.toFixed(2)}`);
  console.log(`      • Used: $${status.used.toFixed(2)}`);
  console.log(`      • Percentage Used: ${status.percentageUsed.toFixed(1)}%`);
  console.log(`      • Remaining: $${status.remaining.toFixed(2)}`);
  console.log(`      • Is Warning: ${status.isWarning}`);
  console.log(`      • Is Exhausted: ${status.isExhausted}`);
  console.log(`      • Message: ${status.message}`);
}

function displayCostAnalytics(analytics: CostAnalytics): void {
  console.log(`      • Period: ${analytics.periodStart} to ${analytics.periodEnd}`);
  console.log(`      • Total Cost: $${analytics.totalCost.toFixed(2)}`);
  console.log(`      • Total Tokens: ${analytics.totalTokens}`);
  console.log(`      • Avg Cost/Token: $${analytics.averageCostPerToken.toFixed(6)}`);

  if (Object.keys(analytics.providerBreakdown).length > 0) {
    console.log('      • Provider Breakdown:');
    for (const [provider, cost] of Object.entries(analytics.providerBreakdown)) {
      console.log(`        - ${provider}: $${cost.toFixed(2)}`);
    }
  }

  if (Object.keys(analytics.modelBreakdown).length > 0) {
    console.log('      • Model Breakdown:');
    for (const [model, cost] of Object.entries(analytics.modelBreakdown)) {
      console.log(`        - ${model}: $${cost.toFixed(2)}`);
    }
  }

  if (Object.keys(analytics.dailyCosts).length > 0) {
    console.log('      • Daily Costs (last 5 days):');
    const dailyCosts = Object.entries(analytics.dailyCosts).sort(
      ([dateA], [dateB]) => dateA.localeCompare(dateB),
    );
    for (const [date, cost] of dailyCosts.reverse().slice(0, 5)) {
      console.log(`        - ${date}: $${cost.toFixed(2)}`);
    }
  }
}

// Run the demo
main().catch((error) => {
  console.error('Demo failed:', error);
  process.exit(1);
});

File: sdk/rust/examples/analytics_demo.rs
//! Analytics and Budget Management Demo
//!
//! This example demonstrates the Circuit Breaker SDK's analytics and budget management
//! capabilities, showing how to track costs, set budgets, and monitor spending across
//! different users and projects.

use circuit_breaker_sdk::{Client, Result};
use serde_json::json;
use std::env;
use tokio;

#[tokio::main]
async fn main() -> Result<()> {
    println!("🔍 Circuit Breaker Analytics & Budget Management Demo");
    println!("====================================================");

    // Initialize the client
    let base_url =
        env::var("CIRCUIT_BREAKER_URL").unwrap_or_else(|_| "http://localhost:4000".to_string());
    let api_key = env::var("CIRCUIT_BREAKER_API_KEY").ok();

    let mut client_builder = Client::builder().base_url(&base_url)?;
    if let Some(key) = api_key {
        client_builder = client_builder.api_key(key);
    }
    let client = client_builder.build()?;

    // Test connection
    match client.ping().await {
        Ok(ping) => println!("✅ Connected to Circuit Breaker server: {}", ping.message),
        Err(e) => {
            println!("❌ Failed to connect to server: {}", e);
            println!(
                "   Make sure the Circuit Breaker server is running at {}",
                base_url
            );
            return Ok(());
        }
    }

    println!("\n📊 Analytics & Budget Management Features:");
    println!("==========================================");

    // 1. Budget Management Demo
    println!("\n1. 💰 Budget Management");
    println!("   ---------------------");

    let user_id = "demo_user_123";
    let project_id = "demo_project_456";

    // Set a monthly budget for a user
    println!("   Setting monthly budget for user: {}", user_id);
    match client
        .analytics()
        .set_budget()
        .user_id(user_id)
        .limit(100.0)
        .period("monthly")
        .warning_threshold(0.8)
        .execute()
        .await
    {
        Ok(budget) => {
            println!("   ✅ Budget set successfully:");
            println!("      • Budget ID: {}", budget.budget_id);
            println!("      • Limit: ${:.2}", budget.limit);
            println!(
                "      • Used: ${:.2} ({:.1}%)",
                budget.used, budget.percentage_used
            );
            println!("      • Remaining: ${:.2}", budget.remaining);
            println!("      • Status: {}", budget.message);

            if budget.is_warning {
                println!("      ⚠️  Warning: Budget is in warning state!");
            }
            if budget.is_exhausted {
                println!("      🚨 Alert: Budget is exhausted!");
            }
        }
        Err(e) => {
            println!(
                "   ⚠️  Could not set budget (server may not be running): {}",
                e
            );
        }
    }

    // Set a project budget
    println!("\n   Setting project budget for: {}", project_id);
    match client
        .analytics()
        .set_budget()
        .project_id(project_id)
        .limit(500.0)
        .period("monthly")
        .warning_threshold(0.75)
        .execute()
        .await
    {
        Ok(budget) => {
            println!("   ✅ Project budget set:");
            println!("      • Limit: ${:.2}", budget.limit);
            println!("      • Warning at: {:.0}%", 0.75 * 100.0);
        }
        Err(e) => {
            println!("   ⚠️  Could not set project budget: {}", e);
        }
    }

    // 2. Budget Status Monitoring
    println!("\n2. 📈 Budget Status Monitoring");
    println!("   ----------------------------");

    // Check user budget status
    println!("   Checking budget status for user: {}", user_id);
    match client
        .analytics()
        .budget_status()
        .user_id(user_id)
        .get()
        .await
    {
        Ok(status) => {
            println!("   ✅ User budget status:");
            display_budget_status(&status);
        }
        Err(e) => {
            println!("   ⚠️  Could not get user budget status: {}", e);
        }
    }

    // Check project budget status
    println!("\n   Checking budget status for project: {}", project_id);
    match client
        .analytics()
        .budget_status()
        .project_id(project_id)
        .get()
        .await
    {
        Ok(status) => {
            println!("   ✅ Project budget status:");
            display_budget_status(&status);
        }
        Err(e) => {
            println!("   ⚠️  Could not get project budget status: {}", e);
        }
    }

    // 3. Cost Analytics
    println!("\n3. 📊 Cost Analytics");
    println!("   -----------------");

    let start_date = "2024-01-01";
    let end_date = "2024-01-31";

    println!(
        "   Getting cost analytics for user: {} ({} to {})",
        user_id, start_date, end_date
    );
    match client
        .analytics()
        .cost_analytics()
        .user_id(user_id)
        .date_range(start_date, end_date)
        .get()
        .await
    {
        Ok(analytics) => {
            println!("   ✅ Cost analytics retrieved:");
            display_cost_analytics(&analytics);
        }
        Err(e) => {
            println!("   ⚠️  Could not get cost analytics: {}", e);
        }
    }

    // Get project analytics
    println!("\n   Getting cost analytics for project: {}", project_id);
    match client
        .analytics()
        .cost_analytics()
        .project_id(project_id)
        .date_range(start_date, end_date)
        .get()
        .await
    {
        Ok(analytics) => {
            println!("   ✅ Project cost analytics:");
            display_cost_analytics(&analytics);
        }
        Err(e) => {
            println!("   ⚠️  Could not get project analytics: {}", e);
        }
    }

    // 4. Convenience Functions Demo
    println!("\n4. 🛠️  Convenience Functions");
    println!("   -------------------------");

    // Using convenience functions
    println!("   Using convenience functions for common operations:");

    // Budget status convenience function
    match circuit_breaker_sdk::budget_status(&client)
        .user_id(user_id)
        .get()
        .await
    {
        Ok(status) => {
            println!(
                "   ✅ Convenience budget status: ${:.2} used of ${:.2}",
                status.used, status.limit
            );
        }
        Err(e) => {
            println!("   ⚠️  Convenience budget status failed: {}", e);
        }
    }

    // Cost analytics convenience function
    match circuit_breaker_sdk::cost_analytics(&client, start_date, end_date)
        .user_id(user_id)
        .get()
        .await
    {
        Ok(analytics) => {
            println!(
                "   ✅ Convenience analytics: ${:.2} total cost",
                analytics.total_cost
            );
        }
        Err(e) => {
            println!("   ⚠️  Convenience analytics failed: {}", e);
        }
    }

    // Set budget convenience function
    match circuit_breaker_sdk::set_budget(&client, 200.0, "monthly")
        .user_id("convenience_user")
        .execute()
        .await
    {
        Ok(budget) => {
            println!("   ✅ Convenience budget set: ${:.2} limit", budget.limit);
        }
        Err(e) => {
            println!("   ⚠️  Convenience budget set failed: {}", e);
        }
    }

    // 5. Real-time Cost Monitoring (Future Feature)
    println!("\n5. ⏰ Real-time Cost Monitoring");
    println!("   -----------------------------");
    println!(
        "   Real-time cost monitoring via subscriptions will be available in a future release."
    );
    println!("   This will allow you to:");
    println!("   • Subscribe to cost updates as they happen");
    println!("   • Get real-time alerts when budgets are exceeded");
    println!("   • Monitor spending patterns in real-time");

    // Demonstrate that subscription isn't implemented yet
    match client
        .analytics()
        .subscribe_cost_updates(Some(user_id))
        .await
    {
        Ok(_) => {
            println!("   ✅ Subscribed to cost updates");
        }
        Err(e) => {
            println!("   ⚠️  Cost update subscriptions: {}", e);
        }
    }

    // 6. Advanced Analytics Scenarios
    println!("\n6. 🔬 Advanced Analytics Scenarios");
    println!("   --------------------------------");

    // Multi-date range analysis
    let date_ranges = vec![
        ("2024-01-01", "2024-01-31", "January"),
        ("2024-02-01", "2024-02-29", "February"),
        ("2024-03-01", "2024-03-31", "March"),
    ];

    println!("   Analyzing costs across multiple months:");
    for (start, end, month) in date_ranges {
        match client
            .analytics()
            .cost_analytics()
            .user_id(user_id)
            .date_range(start, end)
            .get()
            .await
        {
            Ok(analytics) => {
                println!(
                    "   • {}: ${:.2} total, {} tokens used",
                    month, analytics.total_cost, analytics.total_tokens
                );
            }
            Err(_) => {
                println!("   • {}: No data available", month);
            }
        }
    }

    // Budget health check
    println!("\n   Budget Health Check:");
    match client
        .analytics()
        .budget_status()
        .user_id(user_id)
        .get()
        .await
    {
        Ok(status) => {
            let health = if status.is_exhausted {
                "🚨 CRITICAL"
            } else if status.is_warning {
                "⚠️  WARNING"
            } else if status.percentage_used > 50.0 {
                "🟡 MODERATE"
            } else {
                "✅ HEALTHY"
            };

            println!(
                "   Budget Health: {} ({:.1}% used)",
                health, status.percentage_used
            );

            // Recommendations
            if status.is_exhausted {
                println!("   💡 Recommendation: Increase budget limit or optimize usage");
            } else if status.is_warning {
                println!("   💡 Recommendation: Monitor usage closely, consider optimizations");
            } else if status.percentage_used > 50.0 {
                println!(
                    "   💡 Recommendation: Review usage patterns for optimization opportunities"
                );
            } else {
                println!("   💡 Budget is healthy - continue current usage patterns");
            }
        }
        Err(e) => {
            println!("   ⚠️  Could not perform health check: {}", e);
        }
    }

    println!("\n🎉 Analytics Demo Complete!");
    println!("============================");
    println!("This demo showcased:");
    println!("• Budget management for users and projects");
    println!("• Real-time budget status monitoring");
    println!("• Comprehensive cost analytics");
    println!("• Convenience functions for common operations");
    println!("• Advanced analytics scenarios and health checks");
    println!("\nThe Analytics client provides powerful tools for:");
    println!("• Cost control and budget management");
    println!("• Usage optimization and monitoring");
    println!("• Financial planning and reporting");
    println!("• Multi-tenant cost tracking");

    Ok(())
}

fn display_budget_status(status: &circuit_breaker_sdk::BudgetStatus) {
    println!("      • Budget ID: {}", status.budget_id);
    println!("      • Limit: ${:.2}", status.limit);
    println!("      • Used: ${:.2}", status.used);
    println!("      • Percentage Used: {:.1}%", status.percentage_used);
    println!("      • Remaining: ${:.2}", status.remaining);
    println!("      • Is Warning: {}", status.is_warning);
    println!("      • Is Exhausted: {}", status.is_exhausted);
    println!("      • Message: {}", status.message);
}

fn display_cost_analytics(analytics: &circuit_breaker_sdk::CostAnalytics) {
    println!(
        "      • Period: {} to {}",
        analytics.period_start, analytics.period_end
    );
    println!("      • Total Cost: ${:.2}", analytics.total_cost);
    println!("      • Total Tokens: {}", analytics.total_tokens);
    println!(
        "      • Avg Cost/Token: ${:.6}",
        analytics.average_cost_per_token
    );

    if !analytics.provider_breakdown.is_empty() {
        println!("      • Provider Breakdown:");
        for (provider, cost) in &analytics.provider_breakdown {
            println!("        - {}: ${:.2}", provider, cost);
        }
    }

    if !analytics.model_breakdown.is_empty() {
        println!("      • Model Breakdown:");
        for (model, cost) in &analytics.model_breakdown {
            println!("        - {}: ${:.2}", model, cost);
        }
    }

    if !analytics.daily_costs.is_empty() {
        println!("      • Daily Costs (last 5 days):");
        let mut daily_costs: Vec<_> = analytics.daily_costs.iter().collect();
        daily_costs.sort_by(|a, b| a.0.cmp(b.0));
        for (date, cost) in daily_costs.iter().rev().take(5) {
            println!("        - {}: ${:.2}", date, cost);
        }
    }
}
